makeflow/src/mountfile.c

```diff
@@ -0,0 +1,660 @@
+		return -1;
+	}
+
+	if(err_num) return -1;
+
+	return 0;
+}
+
+/* check_cache_dir checks the validity of the cache dir, and create it if possible.
+ * @param cache: a file path.
+ * return 0 on success, return non-zero on failure.
+ */
+int check_cache_dir(const char *cache) {
+	struct stat st;
+
+	/* Checking principle: the cache must locate under the CWD. */
+	if(!cache || !*cache) {
+		LDEBUG("the cache (%s) can not be empty!\n", cache);
+		fprintf(stderr, "the cache (%s) can not be empty!\n", cache);
+		return -1;
+	}
+
+	/* Check whether the cache is an absolute path. */
+	if(cache[0] == '/') {
+		LDEBUG("the cache (%s) should not be an absolute path!\n", cache);
+		fprintf(stderr, "the cache (%s) should not be an absolute path!\n", cache);
+		return -1;
+	}
+
+	/* check whether cache includes .. */
+	if(path_has_doubledots(cache)) {
+		LDEBUG("the cache (%s) include ..!\n", cache);
+		fprintf(stderr, "the cache (%s) include ..!\n", cache);
+		return -1;
+	}
+
+	/* check whether cache includes any symlink link, this check prevent the makeflow breaks out the CWD. */
+	if(path_has_symlink(cache)) {
+		LDEBUG("the cache (%s) should not include any symbolic link!\n", cache);
+		fprintf(stderr, "the cache (%s) should not include any symbolic link!\n", cache);
+		return -1;
+	}
+
+	/* Check whether cache already exists. */
+	if(access(cache, F_OK)) {
+		mode_t default_dirmode = S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH;
+		if(errno != ENOENT) {
+			LDEBUG("access(%s) failed: %s\n", cache, strerror(errno));
+			return -1;
+		}
+
+		LDEBUG("the cache (%s) does not exist, creating it ...\n", cache);
+		if(mkdir_recursive(cache, default_dirmode)) {
+			LDEBUG("mkdir_recursive(%s) failed: %s\n", cache, strerror(errno));
+			return -1;
+		}
+		return 0;
+	}
+
+	/* check whether cache is a dir */
+	if(lstat(cache, &st) == 0) {
+		if(!S_ISDIR(st.st_mode)) {
+			LDEBUG("the cache (%s) should be a dir!\n", cache);
+			fprintf(stderr, "the cache (%s) should be a dir!\n", cache);
+			return -1;
+		}
+	} else {
+		LDEBUG("lstat(%s) failed: %s!\n", cache, strerror(errno));
+		fprintf(stderr, "lstat(%s) failed: %s!\n", cache, strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+int mount_install_all(struct dag *d) {
+	struct list *list;
+	struct dag_file *df;
+	char *cache_dir;
+
+	if(!d) return 0;
+
+	if(d->cache_dir) {
+		/* check the validity of the cache_dir and create it if neccessary and feasible. */
+		if(check_cache_dir(d->cache_dir)) {
+			return -1;
+		}
+
+	} else {
+		/* Create a unique cache dir */
+		cache_dir = xxstrdup(".makeflow_cache.XXXXXX");
+		if(!cache_dir) {
+			LDEBUG("xxstrdup failed: %s\n", strerror(errno));
+			return -1;
+		}
+
+		if(mkdtemp(cache_dir) == NULL) {
+			LDEBUG("mkdtemp(%s) failed: %s\n", cache_dir, strerror(errno));
+			free(cache_dir);
+			return -1;
+		}
+
+		d->cache_dir = cache_dir;
+	}
+
+	/* log the cache dir info */
+	makeflow_log_cache_event(d, d->cache_dir);
```

Comment:
Do we want a new log event every time we run Makeflow with mount or only when the mount changes? @btovar @dthain 

Do repeated runs all contain a new log entry for cache_dir?